src/doctypes/conflicts.js

```diff
@@ -0,0 +1,146 @@
+import mergeWith from 'lodash/mergeWith'
+import mapValues from 'lodash/mapValues'
+import { schema } from './schema'
+
+export const isDocumentUpdateConflict = error => {
+  try {
+    const data = JSON.parse(error.message)
+    return (
+      data.error === 'conflict' ||
+      data.status === '409' ||
+      data.reason === 'Document update conflict.'
+    )
+  } catch (e) {
+    return false
+  }
+}
+
+const DEFAULT_CONFLICT_RESOLUTION_POLICY = 'server'
+
+/**
+ * Zips values contained in valuesByName entries. Each item of the
+ * resulting array is an object keyed by the "name" under which the
+ * value was found.
+ *
+ * @example
+ * ```
+ * > objectZip({
+ *   server: [{ id: 1, name: 'toto'}, { id: 2: name: 'foo'}],
+ *   user: [{ id: 1, name: 'tata'}]
+ * })
+ * [
+ *   {
+ *     server: { id: 1, name: 'toto' },
+ *     user: { id: 1, name: 'tata'}
+ *    },
+ *   { server: { id: 2, name: 'foo'} }
+ * ]
+ * ```
+ */
+const objectZip = (valuesByName, idAttribute) => {
+  const groups = {}
+
+  Object.entries(valuesByName).forEach(([valueName, values]) => {
+    values.forEach(value => {
+      const id = value[idAttribute]
+      groups[id] = groups[id] || {}
+      Object.assign(groups[id], {
+        [valueName]: value
+      })
+    })
+  })
+
+  return Object.values(groups)
+}
+
+const DELETED = {}
+
+const mergeArrays = (serverValues, userValues, schema, merger) => {
+  const groups = objectZip(
+    {
+      server: serverValues,
+      user: userValues
+    },
+    schema.idAttribute
+  )
+
+  return Object.values(groups)
+    .map(group => {
+      if (!group.user) {
+        return DELETED
+      }
+      return mergeWith(group.server, group.user, merger)
+    })
+    .filter(x => x !== DELETED)
+}
+
+const objectCustomizer = customizersByAttr => {
+  return (serverValue, userValue, key) => {
+    const customizer = customizersByAttr[key]
+    if (customizer) {
+      return customizer(serverValue, userValue, key)
+    } else {
+      return conflictResolutionPolicies[DEFAULT_CONFLICT_RESOLUTION_POLICY](
+        serverValue,
+        userValue,
+        key
+      )
+    }
+  }
+}
+
+const conflictResolutionPolicies = {
+  user: (serverValue, userValue) => userValue,
+  server: serverValue => serverValue
+}
+
+const policyCustomizer = policy => {
+  return (
+    conflictResolutionPolicies[policy] ||
+    conflictResolutionPolicies[DEFAULT_CONFLICT_RESOLUTION_POLICY]
+  )
+}
+
+const arrayCustomizer = (schema, valueMerger) => (serverValues, userValues) =>
+  mergeArrays(serverValues, userValues, schema, valueMerger)
+
+const schemaCustomizer = schema => {
+  if (schema.attributes) {
+    const customizer = objectCustomizer(
+      mapValues(schema.attributes, schemaCustomizer)
+    )
+    if (schema.type === 'array') {
+      return arrayCustomizer(schema, customizer)
+    } else {
+      return customizer
+    }
+  } else {
+    const policyFn = policyCustomizer(schema.conflictResolution)
+    return policyFn
+  }
+}
```

Comment:
i find this function and the subfunction it's calling are hard to follow tbh. I don't see anything wrong with them, but mentally it's difficult to see what is happening and what we are working towards. I don't have a suggestion for a rewrite, but I think some jsdoc would be helpful.

Reply:
I reorganized the file, added JSDocs and renamed functions to try and be clearer. Let me know what you think :)

Reply:
I think it's a lot better, thanks! `customizers.js` is still dense, but with the jsdoc everywhere it's already much easier to follow along.